### PR TITLE
Fix OR semantics for SmartAssertions

### LINKED ISSUE #6886: Smart assertions eagerly evaluate rhs of logical expressions
Smart assertions eagerly evaluate rhs of logical expressions, even if it's not logically necessary, so such example assertion fails if `option` is empty:
```Scala
assertTrue(option.isEmpty || option.get == 1)

// fails with
// • Option was None
// option.isEmpty || option.get == 1
// option = None
```
while logically it shouldn't fail.

### PR DESCRIPTION
**Pull request by minedeljkovic**
Attempt to fix #6886 based on [the comment](https://github.com/zio/zio/issues/6886#issuecomment-1142329765) from @adamgfraser.

### CODE REVIEW

File: test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala [hunk 1]
```diff
@@ -71,6 +71,18 @@ object SmartAssertionSpec extends ZIOBaseSpec {
...
       assertTrue(myString.contains("aoseunoth") && myString == "coool")
     } @@ failing,
+    test("or succeeds if lhs side succeeds") {
+      val myOption = Option.empty[Int]
+      assertTrue(myOption.isEmpty || myOption.get == 1)
+    },
+    test("or succeeds if rhs side succeeds") {
+      val myOption = Option.empty[Int]
+      assertTrue(myOption.get == 1 || myOption.isEmpty)
+    },
```
Comment (minedeljkovic) [inline test-tests/shared/src/test/scala/zio/test/SmartAssertionSpec.scala:81]: I'm not sure we want this semantics for rhs, though.
For example, such boolean expression will fail in plain Scala `option.get == 1 || option.isEmpty`.

File: test/shared/src/main/scala/zio/test/Trace.scala [hunk 2]
```diff
@@ -197,7 +197,12 @@ object Trace {
   }
 
   private[test] case class Or(left: Trace[Boolean], right: Trace[Boolean]) extends Trace[Boolean] {
-    override def result: Result[Boolean] = left.result.zipWith(right.result)(_ || _)
+    override def result: Result[Boolean] =
+      (left.result, right.result) match {
+        case (Result.Succeed(true), _) => Result.succeed(true)
+        case (_, Result.Succeed(true)) => Result.succeed(true)
```
Comment (jdegoes) [inline test/shared/src/main/scala/zio/test/Trace.scala:203]: I don't think we want this because it violates the ordinary meaning of `false || true`.

**Review by jdegoes — CHANGES_REQUESTED** [review]
Just one revision, then will be good to merge. Thanks!

**Comment by jdegoes** [comment]
@minedeljkovic Please re-open if you get a chance to make the one suggested revision. Thanks!

**Pull request closed without merging**